### PR TITLE
Dryout address#159

### PR DESCRIPTION
**Pull request by vijayakarumudi**
Closes AgileVentures/MetPlus_tracker#159

I added new "state" attribute to the address table. Added model validation and Rspec tests for that field.
For DRYout address, I created new address form and added that form to branch partial form and updated branch create/edit  and show page.
I also did DRYout company_registrations forms create/edit and companies forms.
Cucumber test for branch create/ edit and company create and duplicate Ein numbers completed successfully.

### CODE REVIEW

File: app/controllers/branches_controller.rb [hunk 1]
```diff
@@ -53,7 +53,7 @@ def destroy
   
   def branch_params
     params.require(:branch).permit(:code, 
-            address_attributes: [:id, :street, :city, :zipcode])
+            address_attributes: [:id, :street, :city, :zipcode,:state])
```
Comment (joaopapereira) [inline app/controllers/branches_controller.rb]: Add a space after the comma

File: app/controllers/companies_controller.rb [hunk 2]
```diff
@@ -36,6 +36,6 @@ def company_params
     :website, :ein, :description,
     company_people_attributes: [:id, :first_name, :last_name, :phone, :email,
                                 :password, :password_confirmation],
-    addresses_attributes: [:id, :street, :city, :zipcode])
+    addresses_attributes: [:id, :street, :city, :zipcode,:state])
```
Comment (joaopapereira) [inline app/controllers/companies_controller.rb]: Add a space after the comma

File: app/controllers/company_registrations_controller.rb [hunk 3]
```diff
@@ -134,12 +134,12 @@ def deny
...
   private
   def company_params
-    params.require(:company).permit(:name, :email, :phone,
+    params.require(:company).permit(:name, :email, :phone, :fax,
     :website, :ein, :description,
     company_people_attributes: [:id, :first_name, :last_name,
                         :phone, :email, :title,
                         :password, :password_confirmation],
-    addresses_attributes: [:id, :street, :city, :zipcode])
+    addresses_attributes: [:id, :street, :city, :zipcode,:state])
```
Comment (joaopapereira) [inline app/controllers/company_registrations_controller.rb]: Add a space after the comma

File: app/views/branches/_form.html.haml [hunk 4]
```diff
@@ -4,24 +4,14 @@
     .col-sm-3
       = branch_form.text_field :code, size: 8, class: 'form-control'
 
-  = branch_form.fields_for :address do |address_form|
-    .form-group
-      = address_form.label :street, class: 'control-label col-sm-2 col-sm-offset-1'
-      .col-sm-5
-        = address_form.text_field :street, class: 'form-control'
+  - @parent_form = branch_form
```
Comment (joaopapereira) [inline app/views/branches/_form.html.haml]: We  could have used the locals to pass this variable like:
render partial: 'address/addressform', locals: {fields_for_name: 'address', address_form: branch_form}

And in the addressForm Partial use
address_form.fields_for fields_for_name.to_sym do |address_form|

This is just a suggestion to make it more readable

**Comment by joaopapereira** [comment]
Looks good, just review my comments and check what is happening with the tests

File: app/controllers/companies_controller.rb (hunk 2 above)
Comment (vijayakarumudi) [inline app/controllers/companies_controller.rb]: Done.

**Comment by vijayakarumudi** [comment]
Last build Task management feature test failed.
I fixed that one. now it is passed.
Now one integration test is called company registration denial failed.

File: app/controllers/branches_controller.rb (hunk 1 above)
Comment (vijayakarumudi) [inline app/controllers/branches_controller.rb]: I fixed it.

File: app/controllers/companies_controller.rb (hunk 2 above)
Comment (vijayakarumudi) [inline app/controllers/companies_controller.rb]: I fixed it

File: app/controllers/company_registrations_controller.rb (hunk 3 above)
Comment (vijayakarumudi) [inline app/controllers/company_registrations_controller.rb]: I added space after the comma.

File: app/views/branches/_form.html.haml (hunk 4 above)
Comment (vijayakarumudi) [inline app/views/branches/_form.html.haml]: I tried your  suggestion. It is only working in the branch level. Variable scope is the problem when we have nested forms like companies and company_registration.

File: app/views/companies/_form.html.haml [hunk 5]
```diff
@@ -1,10 +1,10 @@
 
 = form_for company, url: company_path, method: :patch,                   |
                               html: {class: 'form-horizontal'} do |f|    |
-  - @cmp_form = f
+  - @address_form = f
   = render 'subform'
 
-  .form-group
+    .form-group
```
Comment (joaopapereira) [inline app/views/companies/_form.html.haml]: Is this indentation correct?

File: app/controllers/branches_controller.rb [hunk 6]
```diff
@@ -53,7 +53,7 @@ def destroy
   
   def branch_params
     params.require(:branch).permit(:code, 
-            address_attributes: [:id, :street, :city, :zipcode])
+            address_attributes:        [:id, :street, :city, :zipcode, :state])
```
Comment (joaopapereira) [inline app/controllers/branches_controller.rb]: is the spacing between address_attributes and the [ correct?

Comment (vijayakarumudi) [inline app/controllers/branches_controller.rb]: yes

Comment (vijayakarumudi) [inline app/controllers/branches_controller.rb]: I removed the space

File: app/views/companies/_form.html.haml (hunk 5 above)
Comment (vijayakarumudi) [inline app/views/companies/_form.html.haml]: I updated the indentation

**Comment by joaopapereira** [comment]
Going to merge

**Pull request merged**